picard/util/pipe.py

```diff
@@ -0,0 +1,287 @@
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+
+
+import concurrent.futures
+import os
+from typing import Optional
+
+from picard import PICARD_APP_ID
+from picard.const.sys import (
+    IS_MACOS,
+    IS_WIN,
+)
+
+
+if IS_WIN:
+    import win32pipe  # type: ignore
+    import win32file  # type: ignore
+    from pywintypes import error as WinApiError  # type: ignore
+
+
+class PipeError(Exception):
+    pass
+
+
+class PipeErrorInvalidArgs(PipeError):
+    MESSAGE = "ERROR: Pipe() args argument has to be iterable"
+
+    def __init__(self, msg):
+        super().__init__(f"{self.MESSAGE}: {msg}.")
+
+
+class PipeErrorNotFound(PipeError):
+    MESSAGE = "ERROR: Pipe doesn't exist."
+
+    def __init__(self):
+        super().__init__(self.MESSAGE)
+
+
+class PipeErrorBroken(PipeError):
+    MESSAGE = "ERROR: Pipe is broken."
+
+    def __init__(self):
+        super().__init__(self.MESSAGE)
+
+
+class PipeErrorInvalidResponse(PipeError):
+    MESSAGE = "ERROR: Invalid response from pipe:"
+
+    def __init__(self, response):
+        super().__init__(f"{self.MESSAGE} {response}")
+
+
+class PipeErrorWin(PipeError):
+    MESSAGE = "ERROR: Windows API error\n"
+
+    def __init__(self, winerror):
+        super().__init__(f"{self.MESSAGE}{winerror}")
+
+
+class PipeErrorNoPermission(PipeError):
+    MESSAGE = "ERROR: No permissions for creating a pipe\n"
+
+    def __init__(self, exc):
+        super().__init__(f"{self.MESSAGE}{exc}")
+
+
+class Pipe:
+    NO_RESPONSE_MESSAGE: str = "No response from FIFO"
+    MESSAGE_TO_IGNORE: str = "Ignore this message, just testing the pipe"
+    TIMEOUT_SECS: float = 1.5
+
+    PIPE_WIN_DIR = "\\\\.\\pipe\\"
+    PIPE_MAC_DIR = os.path.join(os.path.expanduser("~/Library/Application Support/"), PICARD_APP_ID)
+    PIPE_UNIX_DIR = os.getenv('XDG_RUNTIME_DIR')
+    PIPE_UNIX_FALLBACK_DIR = os.path.expanduser("~/.config/MusicBrainz/Picard/pipes/")
+
+    def __init__(self, app_name: str, app_version: str, args=None):
+        if args is None:
+            args = tuple()
+        else:
+            try:
+                args = tuple(args)
+            except TypeError as exc:
+                raise PipeErrorInvalidArgs(exc)
+
+        if not args:
+            args = (self.MESSAGE_TO_IGNORE,)
+
+        self.__app_name: str = app_name
+        self.__app_version: str = app_version
+        self.__is_mac: bool = IS_MACOS
+        self.__is_win: bool = IS_WIN
+
+        # named pipe values needed by windows API
+        if self.__is_win:
+            self.__app_version = self.__app_version.replace(".", "-")
```

Comment:
Since `__app_name` and `__app_version` are only used by  `__generate_filename()`, may be just pass `app_name` and `app_version` to it (making this replacement in it too) and drop those.

Reply:
That's what I also considered, I'll do it then.